### PR TITLE
[1.12] Better dependencies for @Mod

### PR DESCRIPTION
**Pull request by Elix-x**
This PR generally improves what you can define in dependencies in your @Mod, including side only dependencies and order independant-required.

-Added ability to define side only dependencies in `@Mod` by adding
`client` for client and `server` for server.
-Added ability to define required, but order independant dependecies
with only `required`.
-Order of dependency instruction matters. For instance
`after-client-required` will make mod load after dependency on server
and client and require it on client.

### CODE REVIEW

**Comment by asiekierka** [comment]
Are all of these strictly necessary? I can see the reason for side-only dependencies, but not order-independent ones.

**Comment by elix-x** [comment]
No, but they are, in some way, side result of code.
Oreder independent ones can be useful when you just want mod to be required, but you don't care about whether it's loaded before yours or after. For example, when using it's classes in world, you don't care about wther or not it should be loaded after your mod while loading.

**Comment by Actuarius** [comment]
@williewillus added labels [Cleanup]

**Comment by cpw** [comment]
I'm not a fan. Sided dependencies should never be necessary.

**Comment by elix-x** [comment]
Why?
For example, if your mod "A" needs another mod "B" that provides some kind of rendering API, there's no need to install or depend on this mod "B" on server.

**Comment by elix-x** [comment]
Anything else?

**Comment by mezz** [comment]
>Order of dependency instruction matters

This seems unnecessarily confusing. I think the format should be more strictly ordered to limit the possibilities, like 
`[required-][after/before][-client/server]`

**Comment by Elix-x** [comment]
That's possible. But they will require writing (almost) similar dependency string in some cases.
For example, if you want to load always load after the mod and on client also require it, you could just do this `after-client-required:themod`. If we define specific order of instructions, for example `[client/server]-[required]-[after/before]`, then for the same result, you'll have to use `server-after:themod;client-required-after:themod`.
Though that's not a big deal.

**Comment by mezz** [comment]
~~Could you do `client-after:themod;server-required:themod` ? I think it's easier to understand that way.~~

**Comment by mezz** [comment]
Yeah I think that adding 2 requirements is easier to understand than combining them into 1.

**Comment by Elix-x** [comment]
If we choose to keep specific order of instructions, it can be any one we want - `[client/server]-[required]-[after/before]` or `[required]-[after/before]-[client/server]` or whatever.

**Comment by mezz** [comment]
I don't have a strong preference, `client-required-after` seems pretty clear.

**Comment by Elix-x** [comment]
Enforced `side-required-order` order.

**Comment by HenryLoenwind** [comment]
Sorry that I'm too lazy to dive into the patch, but does this handle the "I don't care about side and loading order and whether that other mod is installed or not, but if it is there, it must be at least version xxx"?

**Comment by Elix-x** [comment]
@HenryLoenwind Yes. You can ommit all instructions as long as there is at least one.

**Comment by mezz** [comment]
`continue dep;`

The addition of a label makes this reaallly ugly and hard to follow. Is there better way?

**Comment by Elix-x** [comment]
@mezz At this point it means that parse has failed. So we can create method `onParseFailed` which will throw parse exception. This will remove label as well as all `continue`.

**Comment by mezz** [comment]
I think that makes sense, and would be easier to read. A parse failure here should just crash immediately or else modders could have invalid dependency info and not notice.

**Comment by Elix-x** [comment]
Parse failure crashes immediately with details.
Added detailed cause message.
And removed the label.

**Comment by mezz** [comment]
Thanks, it looks very good my opinion!

Can you make the method static and add a junit test (see `net.minecraftforge.fml.test`) that shows it works as intended?
One of the biggest hurdles to getting this accepted is proving that it doesn't break anything, and a thorough test will ensure this feature works well into the future.

The test should at least try every valid combination and check the results. 
It should also show that invalid strings throw `LoaderException`.

**Comment by Elix-x** [comment]
@mezz `parseFailure` or the parser method?

**Comment by mezz** [comment]
I think `computeDependencies` can safely be made static so that it's easier for you to make a unit test for it without creating a `Loader` instance.

**Comment by Elix-x** [comment]
@mezz All instructions can be tested except for side. It requires runtime FML.

**Comment by mezz** [comment]
Can you make a test mod instead, that does basically the same thing that a junit test would do?

**Comment by Elix-x** [comment]
Added JUnit test instead.
Making test mods would require making a bunch of them and would stall forge environment loading each time.

**Comment by mezz** [comment]
You can split it up the tests so that you don't need `try/catch` or `fail()` by using the annotation to say if you expect it to throw an exception: `@Test(expected = LoaderException.class)`.
Tests that throw exceptions that are not expected will fail.
Tests that do not throw the expected exception will fail too.

**Comment by Elix-x** [comment]
@mezz I can. But now it's just tests changes.

**Comment by Elix-x** [comment]
@mezz Can't expect `LoaderException` because `parseFailure` tries to load FML classes which start throwing NPEs.

File: src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+import java.util.Set;
+
+import org.junit.Test;
+
+import net.minecraftforge.fml.common.Loader;
+import net.minecraftforge.fml.common.LoaderException;
+import net.minecraftforge.fml.common.versioning.ArtifactVersion;
+
+public class BetterModDependenciesTest
```
Comment (mezz) [inline src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java]: This should just be called `ModDependenciesTest`

File: src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    public void test()
+    {
+        Loader.computeDependencies(null, null, null, null);
+
+        Loader.computeDependencies("", null, null, null);
+
+        {
+            Set<ArtifactVersion> requirements = new HashSet<ArtifactVersion>();
+            Loader.computeDependencies("required:supermod3000[1.2,)", requirements, null, null);
+            assertFalse(requirements.isEmpty());
```
Comment (mezz) [inline src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java]: This is poor testing, you should check that the length is 1 and the values in the first ArtifactVersion match what you expect.

File: src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        {
+            Set<ArtifactVersion> requirements = new HashSet<ArtifactVersion>();
+            List<ArtifactVersion> dependencies = new ArrayList<ArtifactVersion>();
+            List<ArtifactVersion> dependants = new ArrayList<ArtifactVersion>();
+            Loader.computeDependencies("after:supermod2000[1.3,);required-before:yetanothermod", requirements, dependencies, dependants);
+            assertFalse(requirements.isEmpty());
```
Comment (mezz) [inline src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java]: Please check all the values, not just that it is not empty.

File: src/main/java/net/minecraftforge/fml/common/Loader.java [hunk 4]
```diff
@@ -680,78 +682,111 @@ public ModClassLoader getModClassLoader()
...
+                            parseFailure(dep, "Order instruction has already been defined");
+                        }
+                    }
+                }
+            }
         }
     }
+    
+    private static void parseFailure(String dependencyString, String cause){
+        FMLLog.log(Level.WARN, "Unable to parse dependency string \"%s\", cause - %s", dependencyString, cause);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/Loader.java]: If you remove this FMLLog you may be able to call `parseFailure` from the tests without it having issues, and then you can expect `LoaderException`.
It already throws an exception so the log is probably not necessary.

Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java]: Ok. I will remove it.

File: src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java (hunk 3 above)
Comment (Elix-x) [reply src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java]: This is possible. I can extend it to check values.

File: src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java (hunk 2 above)
Comment (Elix-x) [reply src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java]: This is possible. I can extend it to check values.

File: src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java (hunk 1 above)
Comment (Elix-x) [reply src/test/java/net/minecraftforge/fml/test/BetterModDependenciesTest.java]: Well, PR is named better mod depenendecies, so i reused it. But whatever, i can change.
It's just pure naming it this point

**Comment by Elix-x** [comment]
@mezz There is a bug with versions (which i encountered previously with FEI) but just tracked it down by accident -
Running this part of code
```java
VersionParser.parseVersionReference("supermod3000@[1.2,)").equals(VersionParser.parseVersionReference("supermod3000@[1.2,)"));
```
Will cause following NPE:
```
java.lang.NullPointerException
	at net.minecraftforge.fml.common.versioning.ComparableVersion.compareTo(ComparableVersion.java:480)
	at net.minecraftforge.fml.common.versioning.DefaultArtifactVersion.compareTo(DefaultArtifactVersion.java:55)
	at net.minecraftforge.fml.common.versioning.DefaultArtifactVersion.compareTo(DefaultArtifactVersion.java:1)
	at net.minecraftforge.fml.common.versioning.Restriction.containsVersion(Restriction.java:84)
	at net.minecraftforge.fml.common.versioning.VersionRange.containsVersion(VersionRange.java:522)
	at net.minecraftforge.fml.common.versioning.DefaultArtifactVersion.containsVersion(DefaultArtifactVersion.java:77)
	at net.minecraftforge.fml.common.versioning.DefaultArtifactVersion.equals(DefaultArtifactVersion.java:49)
...
```

Changing version numbers does not help, removing them - does.

Should probably report this as an issue.

While i am at it - `equals` implementation seems weird - it acts as (and uses) containsVersion, which checks bounds inclusion and not equality.

**Comment by Elix-x** [comment]
@mezz Done. I had to use String workaround (`toString` all version ranges and compare them) for checking because of bug i wrote above.

**Comment by mezz** [comment]
Cool, please report an issue so that can get sorted out.

**Comment by mezz** [comment]
In summary, this PR:

- Allows dependencies without "before" or "after" requirements
- Allows server-only or client-only dependencies

All existing behavior is the same as before.

File: src/main/java/net/minecraftforge/fml/common/Loader.java [hunk 5]
```diff
@@ -114,6 +114,8 @@
 public class Loader
 {
     public static final String MC_VERSION = net.minecraftforge.common.ForgeVersion.mcVersion;
+    private static final ImmutableList<String> DEPENDENCYINSTRUCTIONS = ImmutableList.of("client", "server", "required", "before", "after");
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/common/Loader.java:128]: `DEPENDENCY_INSTRUCTIONS`
Although other constant fields' names are not cased properly, I suggest you to make them proper.

File: src/main/java/net/minecraftforge/fml/common/Loader.java [hunk 6]
```diff
@@ -680,78 +682,110 @@ public ModClassLoader getModClassLoader()
         return modClassLoader;
     }
 
-    public void computeDependencies(String dependencyString, Set<ArtifactVersion> requirements, List<ArtifactVersion> dependencies, List<ArtifactVersion> dependants)
+    public static void computeDependencies(String dependencyString, Set<ArtifactVersion> requirements, List<ArtifactVersion> dependencies, List<ArtifactVersion> dependants)
     {
         if (dependencyString == null || dependencyString.length() == 0)
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/common/Loader.java:702]: `Strings.isNullOrEmpty(dependencyString)`

File: src/main/java/net/minecraftforge/fml/common/Loader.java [hunk 7]
```diff
@@ -680,78 +682,110 @@ public ModClassLoader getModClassLoader()
...
+                        {
+                            parseFailure(dep, "Order instruction has already been defined");
+                        }
+                    }
+                }
+            }
         }
     }
+    
+    private static void parseFailure(String dependencyString, String cause){
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/common/Loader.java]: `failParse`?

File: src/main/java/net/minecraftforge/fml/common/Loader.java [hunk 8]
```diff
@@ -680,78 +682,110 @@ public ModClassLoader getModClassLoader()
...
+            if (!DEPENDENCYINSTRUCTIONS.containsAll(instructions))
             {
-                // You can't require everything
-                if (!targetIsAll)
+                parseFailure(dep, String.format("Found invalid instructions. Only %s are allowed.", DEPENDENCYINSTRUCTIONS.toString().replace('[', ' ').replace(']', ' ').trim()));
+            }
+            boolean hasSide = false;
+            boolean hasRequired = false;
+            boolean hasOrder = false;
+            for (String instruction : instructions)
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/common/Loader.java:735]: Can this be represented by bits like java8 spliterator characteristics?

File: src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        Loader.computeDependencies("", null, null, null);
+
+        {
+            Set<ArtifactVersion> requirements = new HashSet<ArtifactVersion>();
+            Loader.computeDependencies("required:supermod3000@[1.2,)", requirements, null, null);
+            assertTrue(equal(Collections2.transform(requirements, Functions.toStringFunction()), Sets.newHashSet("supermod3000@[1.2,)")));
+        }
+
+        {
+            Set<ArtifactVersion> requirements = new HashSet<ArtifactVersion>();
```
Comment (liach) [inline src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java:40]: `Sets.newHashSet()` and `Lists.newArrayList()`.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 5 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:128]: This is possible, but other releated fields were named this way before i got here.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 6 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:702]: I have not changed this line.
But i can.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 8 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:735]: Forge is java 6.

File: src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java (hunk 9 above)
Comment (Elix-x) [reply src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java:40]: Aren't they equivalent?

Comment (liach) [reply src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java:40]: You know, the reason to use such guava functions is that you do not need to write long code for extra generics like `<ArtifactVersion>`. And for consistency.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 8 above)
Comment (liach) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:735]: No, I mean in `Spliterator` from Java 8, there is an integer operated by binary operations. Each bit of that integer represents a feature of the `Spliterator`. Likewise, you can first search for all instructions, and do operation respectively by checking the bits, making the code cleaner (The point is to make the code clean)

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 7 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java]: Field was named this way.

File: src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java (hunk 9 above)
Comment (Elix-x) [reply src/test/java/net/minecraftforge/fml/test/ModDependenciesTest.java:40]: @mezz Does this need @liach's modification?

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 6 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:702]: @mezz Does this need @liach's modification?

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 5 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:128]: @mezz Does this need @liach's modification?

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 8 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:735]: BitMapping?
Possible.
Will it be cleaner?
I am not sure.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 5 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:128]: only if you want to

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 7 above)
Comment (liach) [reply src/main/java/net/minecraftforge/fml/common/Loader.java]: What field? This is a method instead of a field. Its name should be a verb, meaning an action.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 8 above)
Comment (liach) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:735]:
```java
int inst = 0;
for (String instruction : instructions)
{
    if ("client".equals(instruction))
    {
        inst |= INST_CLIENT;
    }
    else if ("server".equals(instruction))
    //etc., gathering inst from instructions
}
if ((inst & INST_CLIENT) == 0)
{
    if (FMLCommonHandler.getSide().isServer()) 
    {
        continue;
    }
}
// other handles
```

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 6 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:702]: with all of these suggestions: Only if you want to, it's not a terrible suggestion since you're rewriting most of this anyway, but it's certainly not required.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 8 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:735]: @liach Do you remember that we have to make sure that order `side-required-order` is respected?
That's what would make it a _bit_ longer.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 6 above)
Comment (Elix-x) [reply src/main/java/net/minecraftforge/fml/common/Loader.java:702]: Ok. I will wait for more important remarks anyways, and if there are any, i may change some of these.

**Comment by Elix-x** [comment]
Any updates?

**Comment by mezz** [comment]
Waiting on response from @cpw. Since he's rewriting FML he hasn't been keen on changes, but I think this would be useful in the meantime (at least the dependencies without "before" or "after" requirements). I still get occasional bug reports about circular dependencies because someone said "after" instead of "before" somewhere, when the order really doesn't matter to them either way.

**Comment by Elix-x** [comment]
Retargetted to 1.12.
Still no updates?

**Comment by mezz** [comment]
@Elix-x can you update this to resolve the merge conflict? I'll do some final testing then we should be able to merge it.

File: src/main/java/net/minecraftforge/fml/common/Loader.java (hunk 7 above)
Comment (liach) [reply src/main/java/net/minecraftforge/fml/common/Loader.java]: Fix you braces!

**Comment by mezz** [comment]
Taking a closer look at this again, the testing is very limited, and the code is written in a way that is still pretty confusing.
I don't think it supports the case Henry asked about, which is "if __ mod is present, it must meet these version requirements" without any other restrictions.
I am going to use this commit as a starting point and get working on it.

**Comment by mezz** [comment]
I updated this PR, here: https://github.com/MinecraftForge/MinecraftForge/pull/4403
Please review when you have a chance so we can finally get this feature in.
